hbase-server/src/main/java/org/apache/hadoop/hbase/master/HbckChecker.java

```diff
@@ -0,0 +1,220 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.master;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.ScheduledChore;
+import org.apache.hadoop.hbase.ServerName;
+import org.apache.hadoop.hbase.client.RegionInfo;
+import org.apache.hadoop.hbase.util.FSUtils;
+import org.apache.hadoop.hbase.util.HbckRegionInfo;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.yetus.audience.InterfaceStability;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.common.collect.Lists;
+
+/**
+ * Used to do the hbck checking job at master side.
+ */
+@InterfaceAudience.Private
+@InterfaceStability.Evolving
+public class HbckChecker extends ScheduledChore {
+  private static final Logger LOG = LoggerFactory.getLogger(HbckChecker.class.getName());
+
+  private static final String HBCK_CHECKER_INTERVAL = "hbase.master.hbck.checker.interval";
+  private static final int DEFAULT_HBCK_CHECKER_INTERVAL = 60 * 60 * 1000;
+
+  private final MasterServices master;
+
+  /**
+   * This map contains the state of all hbck items.  It maps from encoded region
+   * name to HbckRegionInfo structure.  The information contained in HbckRegionInfo is used
+   * to detect and correct consistency (hdfs/meta/deployment) problems.
+   */
+  private final Map<String, HbckRegionInfo> regionInfoMap = new HashMap<>();
+
+  private final Map<String, ServerName> orphanRegionsOnRS = new HashMap<>();
+  private final List<String> orphanRegionsOnFS = new LinkedList<>();
+  private final Map<String, Pair<ServerName, List<ServerName>>> inconsistentRegions =
+      new HashMap<>();
+
+  private final Map<String, ServerName> orphanRegionsOnRSSnapshot = new HashMap<>();
+  private final List<String> orphanRegionsOnFSSnapshot = new LinkedList<>();
+  private final Map<String, Pair<ServerName, List<ServerName>>> inconsistentRegionsSnapshot =
+      new HashMap<>();
+
+  private volatile boolean running = false;
+
+  public HbckChecker(MasterServices master) {
+    super("HbckChecker-", master,
+        master.getConfiguration().getInt(HBCK_CHECKER_INTERVAL, DEFAULT_HBCK_CHECKER_INTERVAL));
+    this.master = master;
+  }
+
+  @Override
+  protected void chore() {
+    running = true;
+    regionInfoMap.clear();
+    orphanRegionsOnRS.clear();
+    orphanRegionsOnFS.clear();
+    inconsistentRegions.clear();
```

Comment:
We need these clears? Below before populating them, we do clears too.

Reply:
Need to clear first. Then generate this round's HBCK result. Then save it to "snapshot".